### PR TITLE
feat: filters

### LINKED ISSUE #207: Add a generic way to filter
Multiples issues are related to the need of generics filters.

In fact, we want the user to be able to apply filter on any column. Some column will be known but options from the tasksOptions is just a map<string, string>
Also, how do I can get the schema of an item?

### LINKED ISSUE #257: Create new filters
filters: [ (|)
	[{key: , value: }], (&)
	[{key: , value: }],
	[{key: , value: }],
]

avec un oneof sur des messages pour avoir les columns qu'on peut avoir (ça va poser une question sur les _after et _before)

### LINKED ISSUE #303: Use a like and a wildcard
Add it in documentation (comments in `proto` files) and implements it in https://github.com/aneoconsulting/ArmoniK.Core

### LINKED ISSUE #348: Add the ability to filter count-tasks-by-status
This must be done after #349 is merged since we will usage the same filters.

### PR DESCRIPTION
**Pull request by esoubiran-aneo**
Add filters to tasks to try implementation.

fix https://github.com/aneoconsulting/ArmoniK/issues/570
fix #348
fix #303
fix #257
fix #207

### CODE REVIEW

File: Protos/V1/tasks_common.proto [hunk 1]
```diff
@@ -171,27 +105,7 @@ message ListTasksRequest {
...
   int32 page_size = 2; /** The page size. */
 
-  /**
-   * Represents a filter with every available fields.
-   *
-   * Every field can be used as the same time.
-   */
-  message Filter {
-    string session_id = 1; /** The session ID of related tasks. */
-    repeated task_status.TaskStatus status = 2; /** Use the differents status of a task to filter tasks having one of the requested status. */
```
Comment (aneojgurhem) [inline Protos/V1/tasks_common.proto:181]: how do we filter on statuses ?

File: Protos/V1/filters_common.proto [hunk 2]
```diff
@@ -0,0 +1,46 @@
+syntax = "proto3";
+
+package armonik.api.grpc.v1;
+
+option csharp_namespace = "Armonik.Api.gRPC.V1.Tasks";
```
Comment (aneojgurhem) [inline Protos/V1/filters_common.proto:5]: I think this should not be in the `Tasks` namespace

**Comment by aneojgurhem** [comment]
I also think we should remove the messages that we do not use anymore

File: Protos/V1/filters_common.proto (hunk 2 above)
Comment (esoubiran-aneo) [inline Protos/V1/filters_common.proto:5]:
```suggestion
option csharp_namespace = "Armonik.Api.gRPC.V1";
```

Comment (aneojgurhem) [reply Protos/V1/filters_common.proto]: I's put a `K` on `ArmoniK`

File: packages/angular/projects/aneoconsultingfr/armonik.api.angular/src/lib/index.ts [hunk 3]
```diff
@@ -63,24 +104,57 @@ export {
...
+  FilterStatus as TaskFilterStatus,
+  Filters as TaskFilters,
+  FiltersAnd as TaskFiltersAnd,
+  FiltersOr as TaskFiltersOr,
+} from "./generated/tasks-filters.pb";
+export {
+  FilterArrayRange,
+  FilterDateRange,
+  FilterNumberRange,
+  FilterStringRange,
```
Comment (esoubiran-aneo) [inline packages/angular/projects/aneoconsultingfr/armonik.api.angular/src/lib/index.ts:132]:
```suggestion
  FilterStringRange,
  FilterStatusRange,
```

**Comment by esoubiran-aneo** [comment]
We could rename `range` to `operator`.

File: Protos/V1/applications_common.proto [hunk 4]
```diff
@@ -55,23 +33,7 @@ message ListApplicationsRequest {
...
-    string namespace = 3; /** Application namespace. */
-    string service = 4; /** Application service. */
-  }
-
-  /** The filter.
-   *
-   * Must be set for every request but allowed to be empty.
-   */
-  Filter filter = 3;
+  Filters filters = 3; /** The filters. */
```
Comment (melflitty) [inline Protos/V1/applications_common.proto:36]: Why 3?

Comment (aneojgurhem) [reply Protos/V1/applications_common.proto:36]: because page_size is 2

Comment (melflitty) [reply Protos/V1/applications_common.proto:36]: Ah, one filter for each item of the page and one for both?

Comment (aneojgurhem) [reply Protos/V1/applications_common.proto:36]: no, it represents the position of the message in grpc serialization process. These numbers must be unique

File: Protos/V1/filters_common.proto [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  FILTER_NUMBER_OPERATOR_GREATER_THAN_OR_EQUAL = 5; /** Greater than or equal */
+  FILTER_NUMBER_OPERATOR_GREATER_THAN = 6; /** Greater than */
+}
+
+enum FilterDateOperator {
+  FILTER_DATE_OPERATOR_UNSPECIFIED = 0; /** Unspecified */
+  FILTER_DATE_OPERATOR_BEFORE = 1; /** Before */
+  FILTER_DATE_OPERATOR_BEFORE_OR_EQUAL = 2; /** Before or equal */
+  FILTER_DATE_OPERATOR_EQUAL = 3; /** Equal */
+  FILTER_DATE_OPERATOR_AFTER_OR_EQUAL = 4; /** After or equal */
```
Comment (ngruelaneo) [inline Protos/V1/filters_common.proto:32]: il manque le `notequal`

File: Protos/V1/filters_common.proto [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  FILTER_DATE_OPERATOR_AFTER = 5; /** After */
```
Comment (ngruelaneo) [inline Protos/V1/filters_common.proto:33]:
```suggestion
  FILTER_DATE_OPERATOR_NOT_EQUAL = 6; /** Not equal */
```

File: Protos/V1/applications_service.proto [hunk 7]
```diff
@@ -17,9 +17,4 @@ service Applications {
    * Get a applications list using pagination, filters and sorting;
    */
   rpc ListApplications(ListApplicationsRequest) returns (ListApplicationsResponse) {}
-
-  /**
-   * Get count from tasks by status using an application.
-   */
-  rpc CountTasksByStatus(CountTasksByStatusRequest) returns (CountTasksByStatusResponse) {}
```
Comment (aneojgurhem) [inline Protos/V1/applications_service.proto:24]: Did you remove the unused messages ?

Comment (esoubiran-aneo) [reply Protos/V1/applications_service.proto:24]: yes

**Review by aneojgurhem — APPROVED** [review]

File: Protos/V1/sessions_filters.proto [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+message FilterField {
+  oneof filter {
+    FilterString string = 1;
+    FilterNumber number = 2;
+    FilterDate date = 3;
+    FilterBoolean boolean = 4;
+    FilterArray array = 5;
+    FilterStatus status = 6;
+  }
+}
```
Comment (lemaitre-aneo) [inline Protos/V1/sessions_filters.proto:56]: Sorry I'm a bit late to the party, but I think the following could be better.

Like that, you could reuse `FilterString`, `FilterNumber`, `FilterValue`... across all services.

```suggestion

message FilterString {
  string value = 1;
  FilterStringOperator operator = 2;
}

message FilterNumber {
  int64 value = 1;
  FilterNumberOperator operator = 2;
}

message FilterValue {
  oneof condition {
    FilterString string = 1;
    FilterNumber number = 2;
    // ...
  }
}

message FilterField {
  SessionField field = 1;
  FilterValue value_condition;
}
```

Comment (aneojgurhem) [reply Protos/V1/sessions_filters.proto:56]: I don't know how much we'll be able to factorize on my end but let's try !

Comment (esoubiran-aneo) [reply Protos/V1/sessions_filters.proto]: I update proto to support this suggestion. thanks @lemaitre-aneo

File: Protos/V1/filters_common.proto [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+option csharp_namespace = "Armonik.Api.gRPC.V1";
+
+enum FilterStringOperator {
+  FILTER_STRING_OPERATOR_UNSPECIFIED = 0; /** Unspecified */
+  FILTER_STRING_OPERATOR_EQUAL = 1; /** Equal */
+  FILTER_STRING_OPERATOR_NOT_EQUAL = 2; /** Not equal */
+  FILTER_STRING_OPERATOR_CONTAINS = 3; /** Contains */
+  FILTER_STRING_OPERATOR_NOT_CONTAINS = 4; /** Not contains */
+  FILTER_STRING_OPERATOR_STARTS_WITH = 5; /** Starts with */
+  FILTER_STRING_OPERATOR_ENDS_WITH = 6; /** Ends with */
```
Comment (lemaitre-aneo) [inline Protos/V1/filters_common.proto:19]: I would remove `UNSPECIFIED`, and make sure `EQUAL` is with index 0 on all filters.

File: Protos/V1/filters_common.proto [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+enum FilterDateOperator {
+  FILTER_DATE_OPERATOR_UNSPECIFIED = 0; /** Unspecified */
+  FILTER_DATE_OPERATOR_BEFORE = 1; /** Before */
+  FILTER_DATE_OPERATOR_BEFORE_OR_EQUAL = 2; /** Before or equal */
+  FILTER_DATE_OPERATOR_EQUAL = 3; /** Equal */
+  FILTER_DATE_OPERATOR_AFTER_OR_EQUAL = 4; /** After or equal */
+  FILTER_DATE_OPERATOR_AFTER = 5; /** After */
+  FILTER_DATE_OPERATOR_NOT_EQUAL = 6; /** Not equal */
```
Comment (lemaitre-aneo) [inline Protos/V1/filters_common.proto:49]: `NOT_EQUAL` should be grouped with `EQUAL`.

File: Protos/V1/filters_common.proto [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+message FilterStatus {
+  oneof value {
+    result_status.ResultStatus result_status = 1;
+    session_status.SessionStatus session_status = 2;
+    task_status.TaskStatus task_status = 3;
+  }
+  FilterStatusOperator operator = 4;
+}
```
Comment (lemaitre-aneo) [inline Protos/V1/filters_common.proto:81]: This message should disapppear

File: Protos/V1/filters_common.proto (hunk 10 above)
Comment (esoubiran-aneo) [reply Protos/V1/filters_common.proto:49]: yes, will update this.

File: Protos/V1/filters_common.proto (hunk 9 above)
Comment (esoubiran-aneo) [reply Protos/V1/filters_common.proto:19]: agree, will update this.

File: Protos/V1/filters_common.proto (hunk 11 above)
Comment (esoubiran-aneo) [reply Protos/V1/filters_common.proto]: indeed, it's an artifact

**Review by aneojgurhem — APPROVED** [review]

**Review by lemaitre-aneo — DISMISSED** [review]

**Review by aneojgurhem — APPROVED** [review]

**Review by lemaitre-aneo — APPROVED** [review]

**Pull request merged**